hal/lib/athena/AnalogGyro.cpp

```diff
@@ -0,0 +1,216 @@
+#include "HAL/AnalogAccumulator.h"
+#include "HAL/AnalogInput.h"
+#include "handles/IndexedHandleResource.h"
+
+namespace {
+struct AnalogGyro {
+  HalAnalogInputHandle handle;
+  float voltsPerDegreePerSecond;
+  float offset;
+  uint32_t center;
+};
+}
+
+static constexpr uint32_t kOversampleBits = 10;
+static constexpr uint32_t kAverageBits = 0;
+static constexpr float kSamplesPerSecond = 50.0;
+static constexpr float kCalibrationSampleTime = 5.0;
+static constexpr float kDefaultVoltsPerDegreePerSecond = 0.007;
+
+using namespace hal;
+
+static IndexedHandleResource<HalGyroHandle, AnalogGyro, kNumAccumulators,
+                             HalHandleEnum::AnalogGyro>
+    analogGyroHandles;
+
+static void Wait(double seconds) {
+  if (seconds < 0.0) return;
+  std::this_thread::sleep_for(std::chrono::duration<double>(seconds));
+}
+
+extern "C" {
+HalGyroHandle initializeAnalogGyro(HalAnalogInputHandle analog_handle,
+                                   int32_t* status) {
+  if (!isAccumulatorChannel(analog_handle, status)) {
+    if (*status == 0) {
+      *status = PARAMETER_OUT_OF_RANGE;
+    }
+    return HAL_INVALID_HANDLE;
+  }
+
+  // handle known to be correct, so no need to type check
+  int16_t channel = getHandleIndex(analog_handle);
+
+  auto handle = analogGyroHandles.Allocate(channel, status);
+
+  if (*status != 0)
+    return HAL_INVALID_HANDLE;  // failed to allocate. Pass error back.
+
+  // Initialize port structure
+  auto gyro = analogGyroHandles.Get(handle);
+  if (gyro == nullptr) {  // would only error on thread issue
+    *status = HAL_HANDLE_ERROR;
+    return HAL_INVALID_HANDLE;
+  }
+
+  gyro->handle = analog_handle;
+  gyro->voltsPerDegreePerSecond = 0;
+  gyro->offset = 0;
+  gyro->center = 0;
+
+  return handle;
+}
+
+void setupAnalogGyro(HalGyroHandle handle, int32_t* status) {
+  auto gyro = analogGyroHandles.Get(handle);
+  if (gyro == nullptr) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+
+  setAnalogAverageBits(gyro->handle, kAverageBits, status);
+  if (*status != 0) return;
+  setAnalogOversampleBits(gyro->handle, kOversampleBits, status);
+  if (*status != 0) return;
+  float sampleRate =
+      kSamplesPerSecond * (1 << (kAverageBits + kOversampleBits));
+  setAnalogSampleRate(sampleRate, status);
+  if (*status != 0) return;
+  Wait(0.1);
+
+  setAnalogGyroDeadband(handle, 0.0f, status);
+  if (*status != 0) return;
+}
+
+void freeAnalogGyro(HalGyroHandle handle) { analogGyroHandles.Free(handle); }
+
+void setAnalogGyroParameters(HalGyroHandle handle,
+                             float voltsPerDegreePerSecond, float offset,
+                             uint32_t center, int32_t* status) {
+  auto gyro = analogGyroHandles.Get(handle);
+  if (gyro == nullptr) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+
+  gyro->voltsPerDegreePerSecond = voltsPerDegreePerSecond;
+  gyro->offset = offset;
+  gyro->center = center;
+  setAccumulatorCenter(gyro->handle, center, status);
+}
```

Comment:
It did in the constructor which I currently do. Center just sets whats considered no movement, so doesn't need a reset.